output/mistral_fluentbit/mistral_fluentbit.c

```diff
@@ -0,0 +1,553 @@
+ * point run through the linked list of log entries we have seen and send them
+ * to Fluent Bit. Remove each log_entry from the linked list as they are
+ * processed and destroy them.
+ *
+ * No special handling of data block number errors is done beyond the message
+ * logged by the main plug-in framework. Instead this function will simply
+ * attempt to log any data received as normal.
+ *
+ * On error the mistral_shutdown flag is set to true which will cause the
+ * plug-in to exit cleanly.
+ *
+ * Parameters:
+ *   block_num   - The data block number that was sent in the message. Unused.
+ *   block_error - Was an error detected in the block number. May indicate data
+ *                 corruption. Unused.
+ *
+ * Returns:
+ *   void
+ */
+void mistral_received_data_end(uint64_t block_num, bool block_error)
+{
+    UNUSED(block_num);
+    UNUSED(block_error);
+
+    char *data = NULL;
+    struct timeval time_elapsed;
+    mistral_log *log_entry = log_list_head;
+    uint64_t calculated_timeframe = 0;
+
+    size_t date_len = sizeof("YYYY-MM-DD"); /* strftime format = %F */
+    size_t ts_len = sizeof("YYYY-MM-DDTHH:MI:SS"); /* = %FT%T */
+    /*size_t type_len = sizeof(",\"_type\":\"throttle\""); / * Max possible len * / */
+    char strdate[date_len];
+    char strts[ts_len];
+    struct tm utc_time;
+
+    while (log_entry) {
+        /* Calculate local time */
+        if (localtime_r(&log_entry->epoch.tv_sec, &utc_time) == NULL) {
+            mistral_err("Unable to calculate UTC time for log message: %ld\n",
+                        log_entry->epoch.tv_sec);
+            free(data);
+            mistral_shutdown();
+            return;
+        }
+
+        /* This is thread-safe because we access and modify mistral_plugin_start and
+         * mistral_plugin_end
+         * only in this thread. */
+
+        if (!timerisset(&mistral_plugin_end)) {
+            gettimeofday(&mistral_plugin_end, NULL);
+        }
+
+        timersub(&mistral_plugin_end, &mistral_plugin_start, &time_elapsed);
+        calculated_timeframe = 1000000 * time_elapsed.tv_sec + time_elapsed.tv_usec;
+
+        strftime(strdate, date_len, "%F", &utc_time);
+        strftime(strts, ts_len, "%FT%T", &utc_time);
+
+        /* Command and filename must be JSON escaped */
+        char *command = fluentbit_escape(log_entry->command);
+        char *file = fluentbit_escape(log_entry->file);
+        char *path = fluentbit_escape(log_entry->path);
+        const char *job_gid = (log_entry->job_group_id[0] == 0) ? "N/A" : log_entry->job_group_id;
+        const char *job_id = (log_entry->job_id[0] == 0) ? "N/A" : log_entry->job_id;
+        char generic_id[MISTRAL_MAX_BUFFER_SIZE];
+        char *new_data = NULL;
+
+        /* We create the content for the genereric_id. The generic_id is used for queries when the
+         * there is no job id present. The requirement for the generic_id came from Arm.
+         */
+        char *user_name = getenv("USER");
```

Comment:
Doing this repeatedly seems unnecessary, we could get this at plug-in start up.

Reply:
I shall change this in the next release.

Reply:
This is changed now.